aspnetcore/grpc/protobuf.md

```diff
@@ -0,0 +1,401 @@
+    string subject = 1;
+    google.protobuf.Timestamp start = 2;
+    google.protobuf.Duration duration = 3;
+}  
+```
+
+The generated properties in the C# class aren't the .NET date and time types. The properties use the `Timestamp` and `Duration` classes in the `Google.Protobuf.WellKnownTypes` namespace. These classes provide methods for converting to and from `DateTimeOffset`, `DateTime`, and `TimeSpan`.
+
+```csharp
+// Create Timestamp and Duration from .NET DateTimeOffset and TimeSpan
+var meeting = new Meeting
+{
+    Time = Timestamp.FromDateTimeOffset(meetingTime), // also FromDateTime()
+    Duration = Duration.FromTimeSpan(meetingLength)
+};
+
+// Convert Timestamp and Duration to .NET DateTimeOffset and TimeSpan
+var time = meeting.Time.ToDateTimeOffset();
+var duration = meeting.Duration?.ToTimeSpan();
+```
+
+> [!NOTE]
+> The `Timestamp` type works with UTC times. `DateTimeOffset` values always have an offset of zero, and the `DateTime.Kind` property is always `DateTimeKind.Utc`.
+
+### Nullable types
+
+The Protobuf code generation for C# uses the native types, such as `int` for `int32`. So the values are always included and can't be null.
+
+For values that require explicit null, such as using `int?` in your C# code, Protobuf's "Well Known Types" include wrappers that are compiled to nullable C# types. To use them, import `wrappers.proto` into your `.proto` file, like this:
+
+```protobuf  
+syntax = "proto3"
+
+import "google/protobuf/wrappers.proto"
+
+message Person {
+    // ...
+    google.protobuf.Int32Value age = 5;
+}
+```
+
+Protobuf will use .NET nullable types (for example, `int?`) for the generated message property.
+
+The following table shows the complete list of wrapper types with their equivalent C# type:
+
+| C# type   | Well Known Type wrapper       |
+| --------- | ----------------------------- |
+| `bool?`   | `google.protobuf.BoolValue`   |
+| `double?` | `google.protobuf.DoubleValue` |
+| `float?`  | `google.protobuf.FloatValue`  |
+| `int?`    | `google.protobuf.Int32Value`  |
+| `long?`   | `google.protobuf.Int64Value`  |
+| `uint?`   | `google.protobuf.UInt32Value` |
+| `ulong?`  | `google.protobuf.UInt64Value` |
+
+### Decimals
+
+Protobuf doesn't natively support the .NET `decimal` type, just `double` and `float`. There's an ongoing discussion in the Protobuf project about the possibility of adding a standard decimal type to the well-known types, with platform support for languages and frameworks that support it. Nothing has been implemented yet.
+
+It's possible to create a message definition to represent the `decimal` type that would work for safe serialization between .NET clients and servers. But developers on other platforms would have to understand the format being used and implement their own handling for it.
+
+#### Creating a custom decimal type for Protobuf
+
+```protobuf
+package CustomTypes;
+
+// Example: 12345.6789 -> { units = 12345, nanos = 678900000 }
+message Decimal {
```

Comment:
The type name here should be consistent with the `partial` code below. After a few projects I settled on `DecimalValue` as a good name, it doesn't conflict with `System.Decimal` and it's consistent with the Well Known Types naming.